bundles/org.corpus_tools.hexatomic.grid/src/main/java/org/corpus_tools/hexatomic/grid/data/GraphDataProvider.java

```diff
@@ -297,15 +332,66 @@ public void setDataValue(int columnIndex, int rowIndex, Object newValue) {
       return;
     }
-    LabelableElement dataObject = column.getDataObject(rowIndex);
-    if (dataObject instanceof SAnnotation) {
-      SAnnotation anno = (SAnnotation) dataObject;
-      log.debug("Setting value on Salt element '{}' to '{}'.", dataObject, newValue);
-      anno.setValue(newValue);
-    } else {
-      log.debug("Action not implemented: Set text on '{}' to '{}'",
-          dataObject == null ? "NULL" : dataObject.toString());
+    ColumnType columnType = column.getColumnType();
+    SStructuredNode node = column.getDataObject(rowIndex);
+    String annotationQName = column.getColumnValue();
+
+    // If new value is empty, remove annotation
+    if (newValue == null || newValue.equals("")) {
+      // Remove annotation from Salt model
+      if (node != null) {
+        node.removeLabel(column.getColumnValue());
+        log.debug("Removed annotation {} from node {}.", annotationQName, node);
+      }
+      // Remove annotation from all column cells
+      if (node instanceof SToken) {
+        // Annotation can only span a single token, i.e., a single row
+        column.setRow(rowIndex, null);
+      } else if (node instanceof SSpan) {
+        List<SToken> overlappedTokens = graph.getOverlappedTokens(node);
+        for (SToken token : overlappedTokens) {
+          column.setRow(orderedDsTokens.indexOf(token), null);
+        }
+        // If, now, the span has no annotations, remove it
+        if (node.getAnnotations().isEmpty()) {
+          graph.removeNode(node);
+          log.debug("Removed empty span {} from graph {}.", node, graph);
+        }
+      }
+    } else { // Value is not empty or null
+      if (columnType == ColumnType.TOKEN_TEXT) {
+        log.debug("Action not implemented: Set token text");
+      } else if (node instanceof SStructuredNode) {
+        changeAnnotationValue(newValue, column, node);
+      } else if (node == null) { // Span to take annotation doesn't exist yet
+        SToken tokenAtIndex = orderedDsTokens.get(rowIndex);
+        if (columnType == ColumnType.SPAN_ANNOTATION) {
+          // Create new span
+          SSpan span = graph.createSpan(tokenAtIndex);
+          column.setRow(rowIndex, span);
+          log.debug("Created new span {}, spanning token {}.", span, tokenAtIndex);
+          createAnnotation(newValue, columnIndex, span);
+        } else if (columnType == ColumnType.TOKEN_ANNOTATION) {
+          // While the token to attach the annotation is not null in the graph, the node backing up
+          // the cell in the column is, so we need to set it first.
+          column.setRow(rowIndex, tokenAtIndex);
+          // Create new token annotation
+          createAnnotation(newValue, columnIndex, tokenAtIndex);
+        } else {
+          log.debug("Action not implemented: Create token.");
+        }
+      } else {
+        log.debug("Action not implemented: Set text on '{}' to '{}'.", node.toString());
```

Comment:
The second `{}` is not backed by an argument to `debug`. Did you mean to give `newValue` as argument?

```suggestion
        log.debug("Action not implemented: Set text on '{}' to '{}'.", node.toString(), newValue);
```

Reply:
:+1: